### DISCUSSION TITLE
Cyclic reference is `Graph` objects

### DISCUSSION THREAD

**Discussion by zhqrbitee** [networkx/networkx#7697] [Q&A]
My teams recently discovered that some `Graph` objects is not cleaned after they go out of scope and this is related to cyclic reference in `Graph` class due to `edges` and `degree` cached property.
```python
    @cached_property
    def edges(self):
        return EdgeView(self)

    @cached_property
    def degree(self):
        return DegreeView(self)
```

The `cached_property` will cache the `EdgeView`/`DegreeView` in the `__dict__` of graph and those views will in turn keep reference to graph itself.  
A similar issue is found: https://github.com/networkx/networkx/issues/2885 (PR: https://github.com/networkx/networkx/pull/3011). It seems at that issue, we remove the cache of `EdgeView`/`DegreeView` on `__dict__` of graph, but now the equivalent things come back due to use of `cached_property` instead of `property `. 

Since python declare it supports detect cyclic garbage collection, so not sure if this is a very big issue, but I'm wondering:
* why we switch to `cached_property`? Is there a performance issue if we create `EdgeView`/`DegreeView` from scratch every time? 
* could we use weakref in `EdgeView`/`DegreeView` to avoid the cyclic reference?

**Comment by dschult** [comment]
Can you explain the trouble that this causes you?
The cached_property is used instead of property because that we we don't create the same object over and over again. Python's garbage collector should be able to see that it is a cycle dependence and still clean it with the garbage when appropriate. Are ther troubles you are running into?

**Reply by zhqrbitee** [reply]
Hi @dschult , we didn't find it to be a real issue, we just find this when debug memory leak in our application. It turns out to be other issues from C extension (It just makes things harder when debugging those issues).  

>  Python's garbage collector should be able to see that it is a cycle dependence and still clean it with the garbage when appropriate

This is why I didn't open a bug report and start a discussion here. Personally I would try to avoid cycle dependence, but this is a valid argument.

**Reply by rossbar** [reply]
Thanks for sharing the context @zhqrbitee .

I suspect the use of `cached_property` does have significant implications - see e.g. #5600 (it might be nice to add explicit benchmarks that attempt to capture these)! AIUI the cyclic garbage collection is a language feature, so I don't see any problem with relying on it.
